{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "The purpose of this .ipynb is to generate motif-pair in-silico perturbations across canonical motifs and their variants. This will allow us to assess whether motif pairs have preferential distances or spacings between them.\n",
    "\n",
    "# Computational setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n",
      "2021-07-16 15:53:47,190 [WARNING] git-lfs not installed\n",
      "2021-07-16 15:53:48,749 [INFO] Note: detected 80 virtual cores but NumExpr set to maximum of 64, check \"NUMEXPR_MAX_THREADS\" environment variable.\n",
      "2021-07-16 15:53:48,751 [INFO] Note: NumExpr detected 80 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "2021-07-16 15:53:48,752 [INFO] NumExpr defaulting to 8 threads.\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from tensorflow.python.util import deprecation\n",
    "deprecation._PRINT_DEPRECATION_WARNINGS = False\n",
    "\n",
    "#Packages\n",
    "import os\n",
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from pybedtools import BedTool\n",
    "\n",
    "# Settings\n",
    "os.chdir(f'/n/projects/mw2098/collaboration/for_srivastava/analysis/CM_vs_MN_merged_bias_exp_peaks/')\n",
    "pd.set_option('display.max_columns', 100)\n",
    "\n",
    "# Custom commands\n",
    "sys.path.insert(0, f'/n/projects/mw2098/shared_code/bpnet/scripts')\n",
    "from data_format_functions import myround, myfloor, myceiling\n",
    "\n",
    "# function to return key for any value \n",
    "def get_key(val, my_dict): \n",
    "    for key, value in my_dict.items(): \n",
    "        if val == value: \n",
    "            return key \n",
    "    return \"key doesn't exist\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pre-existing variables\n",
    "fasta_file = f'/n/projects/mw2098/genomes/hg19/hg19.fa'\n",
    "model_prefix = 'seq_width1000-lr0.001-lambda100-n_dil_layers9-conv_kernel_size7-tconv_kernel_size7-filters64'\n",
    "prefix = 'CM_vs_MN_merged_bias_exp_peaks'\n",
    "tasks = ['I_WT_D6CM','I_WT_S3MN']\n",
    "\n",
    "# Determined variables\n",
    "model_dir = f'../../models/{prefix}/{model_prefix}'\n",
    "modisco_dir = f'../../modisco/{prefix}/{model_prefix}'\n",
    "\n",
    "\n",
    "motif_seqs = {'NKX2.5': 'TTAAGTG',\n",
    "              'NKX2.5-alt': 'CTAAGTG',\n",
    "              'GATA': 'AGATAAG',\n",
    "              'ISL1': 'CTAATGG',\n",
    "              'LHX': 'TAATTCAATTA',\n",
    "              'NeuroD': 'CAGCTG',\n",
    "              'EBF1': 'TCCCCAGGGA',\n",
    "              'Onecut2': 'TCCCCAGGGA'\n",
    "             }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p ../../figures/{prefix}/8_in_silico_distances/sum\n",
    "!mkdir -p tsv/insilico_perturb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define motif pairs\n",
    "\n",
    "Here, we will look at each homotypic and heterotypic motif pair."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import permutations \n",
    "motif_perms = list(permutations(motif_seqs.keys(), 2))\n",
    "motif_perms.extend([(i,i) for i in motif_seqs.keys()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import BPNet model\n",
    "\n",
    "Load the BPNet model that was trained."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TF-MoDISco is using the TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "from bpnet.BPNet import BPNetSeqModel\n",
    "from bpnet.utils import create_tf_session\n",
    "create_tf_session('0', .4)\n",
    "bpn = BPNetSeqModel.from_mdir(model_dir)  # wrap SeqModel to BPNetSeqModel to get `sim_pred` method"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Define functions\n",
    "\n",
    "We want to define functions that will allow us to generate simulated, injected versions of motif pairs and format them into pd.dfs that can be plotted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import OrderedDict\n",
    "from bpnet.plot.tracks import filter_tracks\n",
    "from bpnet.utils import flatten, unflatten\n",
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "import pandas as pd\n",
    "from scipy.stats import entropy\n",
    "import random\n",
    "from concise.preprocessing import encodeDNA\n",
    "from bpnet.plot.tracks import plot_tracks\n",
    "from kipoi_utils.data_utils import get_dataset_item, numpy_collate_concat\n",
    "from bpnet.functions import mean, softmax\n",
    "from tqdm import tqdm\n",
    "from kipoi_utils.utils import unique_list\n",
    "import matplotlib.pyplot as plt\n",
    "from bpnet.modisco.core import Seqlet\n",
    "\n",
    "from bpnet.simulate import simmetric_kl\n",
    "\n",
    "#Get results of bpnet.BPNet.sim_pred into a tidy pd.df\n",
    "def sim_pred_to_df(sim_pred_profiles):\n",
    "    dfp_ref = pd.DataFrame()\n",
    "    for k in sim_pred_profiles.keys():\n",
    "        df = pd.DataFrame(sim_pred_profiles[k]).reset_index()\n",
    "        df['task'] = os.path.basename(k)\n",
    "        df.columns = ['position','prediction','task']\n",
    "        dfp_ref = dfp_ref.append(df)\n",
    "    return dfp_ref\n",
    "\n",
    "#Get results of bpnet.simulate.generate_sim into a tidy pd.df\n",
    "def generate_sim_to_df(generate_sim_profiles):\n",
    "    dfp_alt = pd.DataFrame()\n",
    "    for d in range(len(generate_sim_profiles)):\n",
    "        dfp = sim_pred_to_df(generate_sim_profiles[d][1]['profile'])\n",
    "        dfp['distance'] = generate_sim_profiles[d][0]\n",
    "        dfp_alt = dfp_alt.append(dfp)\n",
    "    return dfp_alt\n",
    "\n",
    "def generate_sim(bpnet, central_motif, side_motif, side_distances,\n",
    "                 center_coords=[450, 550], repeat=128, contribution=['count', 'profile'], correct=False):\n",
    "    outl = []\n",
    "    tasks = bpnet.tasks\n",
    "    seqlen = bpnet.input_seqlen()\n",
    "    # ref_preds = sim_pred(model, central_motif)\n",
    "    ref_preds = unflatten(bpnet.sim_pred(central_motif,\n",
    "                                         repeat=repeat,\n",
    "                                         contribution=contribution), \"/\")\n",
    "    none_preds = unflatten(bpnet.sim_pred('', '', [],\n",
    "                                          repeat=repeat, contribution=contribution), \"/\")\n",
    "\n",
    "    alt_profiles = []\n",
    "    for dist in tqdm(side_distances):\n",
    "        # alt_preds = sim_pred(model, central_motif, side_motif, [dist])\n",
    "        alt_preds = unflatten(bpnet.sim_pred(central_motif, side_motif, [dist],\n",
    "                                             repeat=repeat, contribution=contribution), \"/\")\n",
    "        if correct:\n",
    "            # Correct for the 'shoulder' effect\n",
    "            #\n",
    "            # this performs: AB - (B - 0)\n",
    "            # Where:\n",
    "            # - AB: contains both, central and side_motif\n",
    "            # - B : contains only side_motif\n",
    "            # - 0 : doesn't contain any motif\n",
    "            edge_only_preds = unflatten(bpnet.sim_pred('', side_motif, [dist],\n",
    "                                                       repeat=repeat, contribution=contribution), \"/\")\n",
    "\n",
    "            alt_preds_f = flatten(alt_preds, '/')\n",
    "            # ref_preds_f = flatten(ref_preds, '/')\n",
    "            none_preds_f = flatten(none_preds, \"/\")\n",
    "            # substract the other counts\n",
    "            alt_preds = unflatten({k: alt_preds_f[k] - v + none_preds_f[k]\n",
    "                                   for k, v in flatten(edge_only_preds, \"/\").items()}, \"/\")\n",
    "            # ref_preds = unflatten({k: ref_preds_f[k] - v  for k,v in flatten(none_preds, \"/\").items()}, \"/\")\n",
    "        alt_profiles.append((dist, alt_preds))\n",
    "\n",
    "        # This normalizes the score by `A` finally yielding:\n",
    "        # (AB - B + 0) / A\n",
    "        scores = get_scores(ref_preds, alt_preds, tasks, central_motif, seqlen, center_coords)\n",
    "\n",
    "        # compute the distance metrics\n",
    "        for task in bpnet.tasks:\n",
    "            d = scores[task]\n",
    "\n",
    "            # book-keeping\n",
    "            d['task'] = task\n",
    "            d['central_motif'] = central_motif\n",
    "            d['side_motif'] = side_motif\n",
    "            d['position'] = dist\n",
    "            d['distance'] = dist - seqlen // 2\n",
    "\n",
    "            outl.append(d)\n",
    "\n",
    "    return pd.DataFrame(outl), alt_profiles\n",
    "\n",
    "def get_scores(ref_pred, alt_pred, tasks, motif, seqlen, center_coords):\n",
    "    d = {}\n",
    "    cstart, cend = center_coords\n",
    "    for task in tasks:\n",
    "        # profile - use the filtered tracks\n",
    "        d[task] = flatten({\"profile\": profile_sim_metrics(ref_pred['profile'][task][cstart:cend],\n",
    "                                                          alt_pred['profile'][task][cstart:cend])}, \"/\")\n",
    "\n",
    "        # contribution scores - use the central motif region\n",
    "        if 'contrib' in ref_pred:\n",
    "            for contrib_score in ref_pred[\"contrib\"][task]:\n",
    "                contrib, contrib_frac = contrib_sim_metrics(ref_pred[\"contrib\"][task][contrib_score],\n",
    "                                                            alt_pred[\"contrib\"][task][contrib_score], motif, seqlen)\n",
    "                d[task] = {f\"contrib/{contrib_score}\": contrib, f\"contrib/{contrib_score}_frac\": contrib_frac, **d[task]}\n",
    "    return d\n",
    "\n",
    "def profile_sim_metrics(ref, alt, pc=0):\n",
    "    d = {}\n",
    "    d['simmetric_kl'] = simmetric_kl(ref, alt).mean() - simmetric_kl(ref, ref).mean()\n",
    "    d['counts'] = alt.sum()\n",
    "    d['counts_frac'] = (alt.sum() + pc) / (ref.sum() + pc)\n",
    "    d['max'] = alt.max()\n",
    "    d['max_frac'] = (alt.max() + pc) / (ref.max() + pc)\n",
    "\n",
    "    max_idx = np.argmax(ref, axis=0)\n",
    "    d['counts_max_ref'] = alt[max_idx, [0]].sum()\n",
    "    d['counts_max_ref_frac'] = (d['counts_max_ref'] + pc) / (ref[max_idx, [0]].sum() + pc)\n",
    "    return d\n",
    "\n",
    "#Get perturbations and convert into tidy.df (dfs = summary, dfp = profiles)\n",
    "def get_sim(motif_comb, motif_seqs, side_distances = np.arange(505, 700), center_coords = [475, 525]):\n",
    "    dfs, profiles = generate_sim(bpn, central_motif=motif_seqs[motif_comb[0]], side_motif=motif_seqs[motif_comb[1]], \n",
    "                       side_distances=side_distances, center_coords=center_coords, contribution=[], correct=True, repeat = 128)\n",
    "    dfs.central_motif = motif_comb[0]\n",
    "    dfs.side_motif = motif_comb[1]\n",
    "    \n",
    "    #Get reference profiles\n",
    "    profiles_ref = bpn.sim_pred(motif_seqs[motif_comb[0]], repeat = 128)\n",
    "    dfp_ref = sim_pred_to_df(profiles_ref)\n",
    "    dfp_ref['distance'] = 'Reference'\n",
    "\n",
    "    #Get perturbed profiles\n",
    "    dfp_alt = generate_sim_to_df(profiles)\n",
    "\n",
    "    #Combine\n",
    "    dfp = dfp_ref.append(dfp_alt)\n",
    "\n",
    "    return dfs, dfp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate distance-separated motif pair measurements\n",
    "\n",
    "After knocking out the 'side' injected motif, measure the effect on the 'central' motif. Do this for varying distances to investigate the effects of motif pair spacing between each set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 110/110 [00:32<00:00,  3.40it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.39it/s]\n",
      "100%|██████████| 110/110 [00:30<00:00,  3.56it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.36it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.38it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.40it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.37it/s]\n",
      "100%|██████████| 110/110 [00:30<00:00,  3.57it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.36it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.41it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.35it/s]\n",
      "100%|██████████| 110/110 [00:33<00:00,  3.33it/s]\n",
      "100%|██████████| 110/110 [00:29<00:00,  3.71it/s]\n",
      "100%|██████████| 110/110 [00:33<00:00,  3.33it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.34it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.39it/s]\n",
      "100%|██████████| 110/110 [00:33<00:00,  3.32it/s]\n",
      "100%|██████████| 110/110 [00:31<00:00,  3.54it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.35it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.38it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.38it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.34it/s]\n",
      "100%|██████████| 110/110 [00:31<00:00,  3.47it/s]\n",
      "100%|██████████| 110/110 [00:33<00:00,  3.32it/s]\n",
      "100%|██████████| 110/110 [00:33<00:00,  3.32it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.35it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.35it/s]\n",
      "100%|██████████| 110/110 [00:31<00:00,  3.48it/s]\n",
      "100%|██████████| 110/110 [00:31<00:00,  3.44it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.34it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.43it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.39it/s]\n",
      "100%|██████████| 110/110 [00:31<00:00,  3.54it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.37it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.41it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.36it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.35it/s]\n",
      "100%|██████████| 110/110 [00:30<00:00,  3.55it/s]\n",
      "100%|██████████| 110/110 [00:33<00:00,  3.33it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.40it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.37it/s]\n",
      "100%|██████████| 110/110 [00:33<00:00,  3.33it/s]\n",
      "100%|██████████| 110/110 [00:30<00:00,  3.55it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.37it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.41it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.40it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.37it/s]\n",
      "100%|██████████| 110/110 [00:30<00:00,  3.58it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.38it/s]\n",
      "100%|██████████| 110/110 [00:31<00:00,  3.45it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.41it/s]\n",
      "100%|██████████| 110/110 [00:31<00:00,  3.47it/s]\n",
      "100%|██████████| 110/110 [00:30<00:00,  3.63it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.41it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.40it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.35it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.40it/s]\n",
      "100%|██████████| 110/110 [00:30<00:00,  3.61it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.35it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.38it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.38it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.36it/s]\n",
      "100%|██████████| 110/110 [00:31<00:00,  3.50it/s]\n",
      "100%|██████████| 110/110 [00:32<00:00,  3.34it/s]\n"
     ]
    }
   ],
   "source": [
    "#%%script false --no-raise-error\n",
    "dfs_all = pd.DataFrame()\n",
    "for perm in motif_perms:\n",
    "    dfs, _ = get_sim(motif_seqs = motif_seqs, motif_comb=perm, side_distances = np.arange(510, 620))\n",
    "    dfs_all = dfs_all.append(dfs)\n",
    "dfs_all.to_csv(f'tsv/insilico_perturb/insilico_summaries.tsv.gz', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in the summaries after they are generated and assign motif pair names and information for simultaneous plotting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>profile/simmetric_kl</th>\n",
       "      <th>profile/counts</th>\n",
       "      <th>profile/counts_frac</th>\n",
       "      <th>profile/max</th>\n",
       "      <th>profile/max_frac</th>\n",
       "      <th>profile/counts_max_ref</th>\n",
       "      <th>profile/counts_max_ref_frac</th>\n",
       "      <th>task</th>\n",
       "      <th>central_motif</th>\n",
       "      <th>side_motif</th>\n",
       "      <th>position</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000289</td>\n",
       "      <td>0.206230</td>\n",
       "      <td>1.164459</td>\n",
       "      <td>0.004500</td>\n",
       "      <td>1.212017</td>\n",
       "      <td>0.004500</td>\n",
       "      <td>1.212017</td>\n",
       "      <td>I_WT_D6CM</td>\n",
       "      <td>NKX2.5</td>\n",
       "      <td>NKX2.5-alt</td>\n",
       "      <td>510</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000415</td>\n",
       "      <td>0.077576</td>\n",
       "      <td>0.968362</td>\n",
       "      <td>0.001631</td>\n",
       "      <td>0.974014</td>\n",
       "      <td>0.001591</td>\n",
       "      <td>0.950136</td>\n",
       "      <td>I_WT_S3MN</td>\n",
       "      <td>NKX2.5</td>\n",
       "      <td>NKX2.5-alt</td>\n",
       "      <td>510</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000478</td>\n",
       "      <td>0.247855</td>\n",
       "      <td>1.399492</td>\n",
       "      <td>0.005421</td>\n",
       "      <td>1.460272</td>\n",
       "      <td>0.005417</td>\n",
       "      <td>1.458996</td>\n",
       "      <td>I_WT_D6CM</td>\n",
       "      <td>NKX2.5</td>\n",
       "      <td>NKX2.5-alt</td>\n",
       "      <td>511</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.000268</td>\n",
       "      <td>0.076581</td>\n",
       "      <td>0.955940</td>\n",
       "      <td>0.001590</td>\n",
       "      <td>0.949813</td>\n",
       "      <td>0.001583</td>\n",
       "      <td>0.945259</td>\n",
       "      <td>I_WT_S3MN</td>\n",
       "      <td>NKX2.5</td>\n",
       "      <td>NKX2.5-alt</td>\n",
       "      <td>511</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0.000214</td>\n",
       "      <td>0.245519</td>\n",
       "      <td>1.386302</td>\n",
       "      <td>0.005375</td>\n",
       "      <td>1.447714</td>\n",
       "      <td>0.005375</td>\n",
       "      <td>1.447714</td>\n",
       "      <td>I_WT_D6CM</td>\n",
       "      <td>NKX2.5</td>\n",
       "      <td>NKX2.5-alt</td>\n",
       "      <td>512</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  profile/simmetric_kl  profile/counts  profile/counts_frac  \\\n",
       "0           0              0.000289        0.206230             1.164459   \n",
       "1           1              0.000415        0.077576             0.968362   \n",
       "2           2              0.000478        0.247855             1.399492   \n",
       "3           3              0.000268        0.076581             0.955940   \n",
       "4           4              0.000214        0.245519             1.386302   \n",
       "\n",
       "   profile/max  profile/max_frac  profile/counts_max_ref  \\\n",
       "0     0.004500          1.212017                0.004500   \n",
       "1     0.001631          0.974014                0.001591   \n",
       "2     0.005421          1.460272                0.005417   \n",
       "3     0.001590          0.949813                0.001583   \n",
       "4     0.005375          1.447714                0.005375   \n",
       "\n",
       "   profile/counts_max_ref_frac       task central_motif  side_motif  position  \\\n",
       "0                     1.212017  I_WT_D6CM        NKX2.5  NKX2.5-alt       510   \n",
       "1                     0.950136  I_WT_S3MN        NKX2.5  NKX2.5-alt       510   \n",
       "2                     1.458996  I_WT_D6CM        NKX2.5  NKX2.5-alt       511   \n",
       "3                     0.945259  I_WT_S3MN        NKX2.5  NKX2.5-alt       511   \n",
       "4                     1.447714  I_WT_D6CM        NKX2.5  NKX2.5-alt       512   \n",
       "\n",
       "   distance  \n",
       "0        10  \n",
       "1        10  \n",
       "2        11  \n",
       "3        11  \n",
       "4        12  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_all = pd.read_csv(f'tsv/insilico_perturb/insilico_summaries.tsv.gz', sep = '\\t')\n",
    "dfs_all.head(n=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>profile/simmetric_kl</th>\n",
       "      <th>profile/counts</th>\n",
       "      <th>profile/counts_frac</th>\n",
       "      <th>profile/max</th>\n",
       "      <th>profile/max_frac</th>\n",
       "      <th>profile/counts_max_ref</th>\n",
       "      <th>profile/counts_max_ref_frac</th>\n",
       "      <th>task</th>\n",
       "      <th>central_motif</th>\n",
       "      <th>side_motif</th>\n",
       "      <th>position</th>\n",
       "      <th>distance</th>\n",
       "      <th>motif_pair</th>\n",
       "      <th>motif_pair_raw</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000289</td>\n",
       "      <td>0.206230</td>\n",
       "      <td>1.164459</td>\n",
       "      <td>0.004500</td>\n",
       "      <td>1.212017</td>\n",
       "      <td>0.004500</td>\n",
       "      <td>1.212017</td>\n",
       "      <td>I_WT_D6CM</td>\n",
       "      <td>NKX2.5</td>\n",
       "      <td>NKX2.5-alt</td>\n",
       "      <td>510</td>\n",
       "      <td>10</td>\n",
       "      <td>NKX2.5_NKX2.5-alt</td>\n",
       "      <td>NKX2.5_NKX2.5-alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000415</td>\n",
       "      <td>0.077576</td>\n",
       "      <td>0.968362</td>\n",
       "      <td>0.001631</td>\n",
       "      <td>0.974014</td>\n",
       "      <td>0.001591</td>\n",
       "      <td>0.950136</td>\n",
       "      <td>I_WT_S3MN</td>\n",
       "      <td>NKX2.5</td>\n",
       "      <td>NKX2.5-alt</td>\n",
       "      <td>510</td>\n",
       "      <td>10</td>\n",
       "      <td>NKX2.5_NKX2.5-alt</td>\n",
       "      <td>NKX2.5_NKX2.5-alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000478</td>\n",
       "      <td>0.247855</td>\n",
       "      <td>1.399492</td>\n",
       "      <td>0.005421</td>\n",
       "      <td>1.460272</td>\n",
       "      <td>0.005417</td>\n",
       "      <td>1.458996</td>\n",
       "      <td>I_WT_D6CM</td>\n",
       "      <td>NKX2.5</td>\n",
       "      <td>NKX2.5-alt</td>\n",
       "      <td>511</td>\n",
       "      <td>11</td>\n",
       "      <td>NKX2.5_NKX2.5-alt</td>\n",
       "      <td>NKX2.5_NKX2.5-alt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.000268</td>\n",
       "      <td>0.076581</td>\n",
       "      <td>0.955940</td>\n",
       "      <td>0.001590</td>\n",
       "      <td>0.949813</td>\n",
       "      <td>0.001583</td>\n",
       "      <td>0.945259</td>\n",
       "      <td>I_WT_S3MN</td>\n",
       "      <td>NKX2.5</td>\n",
       "      <td>NKX2.5-alt</td>\n",
       "      <td>511</td>\n",
       "      <td>11</td>\n",
       "      <td>NKX2.5_NKX2.5-alt</td>\n",
       "      <td>NKX2.5_NKX2.5-alt</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  profile/simmetric_kl  profile/counts  profile/counts_frac  \\\n",
       "0           0              0.000289        0.206230             1.164459   \n",
       "1           1              0.000415        0.077576             0.968362   \n",
       "2           2              0.000478        0.247855             1.399492   \n",
       "3           3              0.000268        0.076581             0.955940   \n",
       "\n",
       "   profile/max  profile/max_frac  profile/counts_max_ref  \\\n",
       "0     0.004500          1.212017                0.004500   \n",
       "1     0.001631          0.974014                0.001591   \n",
       "2     0.005421          1.460272                0.005417   \n",
       "3     0.001590          0.949813                0.001583   \n",
       "\n",
       "   profile/counts_max_ref_frac       task central_motif  side_motif  position  \\\n",
       "0                     1.212017  I_WT_D6CM        NKX2.5  NKX2.5-alt       510   \n",
       "1                     0.950136  I_WT_S3MN        NKX2.5  NKX2.5-alt       510   \n",
       "2                     1.458996  I_WT_D6CM        NKX2.5  NKX2.5-alt       511   \n",
       "3                     0.945259  I_WT_S3MN        NKX2.5  NKX2.5-alt       511   \n",
       "\n",
       "   distance         motif_pair     motif_pair_raw  \n",
       "0        10  NKX2.5_NKX2.5-alt  NKX2.5_NKX2.5-alt  \n",
       "1        10  NKX2.5_NKX2.5-alt  NKX2.5_NKX2.5-alt  \n",
       "2        11  NKX2.5_NKX2.5-alt  NKX2.5_NKX2.5-alt  \n",
       "3        11  NKX2.5_NKX2.5-alt  NKX2.5_NKX2.5-alt  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Assign motif_pair names for faceting\n",
    "motif_pair = []\n",
    "for i in range(dfs_all.shape[0]):\n",
    "    if dfs_all.central_motif.iloc[i] > dfs_all.side_motif.iloc[i]:\n",
    "        val = dfs_all.side_motif.iloc[i] + '_' + dfs_all.central_motif.iloc[i]\n",
    "        motif_pair.append(val)\n",
    "    else:\n",
    "        val = dfs_all.central_motif.iloc[i] + '_' + dfs_all.side_motif.iloc[i]\n",
    "        motif_pair.append(val)\n",
    "dfs_all['motif_pair'] = motif_pair\n",
    "dfs_all['motif_pair_raw'] = dfs_all.central_motif + '_' + dfs_all.side_motif\n",
    "\n",
    "#Make categorigal variable column\n",
    "dfs_all['task'] = pd.Categorical(dfs_all['task'], categories = ['I_WT_D6CM','I_WT_S3MN'], ordered = False)\n",
    "dfs_all.head(n=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the motif pair summaries at their respective distances. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['NKX2.5_NKX2.5-alt', 'GATA_NKX2.5', 'ISL1_NKX2.5', 'LHX_NKX2.5',\n",
       "       'NKX2.5_NeuroD', 'EBF1_NKX2.5', 'NKX2.5_Onecut2',\n",
       "       'GATA_NKX2.5-alt', 'ISL1_NKX2.5-alt', 'LHX_NKX2.5-alt',\n",
       "       'NKX2.5-alt_NeuroD', 'EBF1_NKX2.5-alt', 'NKX2.5-alt_Onecut2',\n",
       "       'GATA_ISL1', 'GATA_LHX', 'GATA_NeuroD', 'EBF1_GATA',\n",
       "       'GATA_Onecut2', 'ISL1_LHX', 'ISL1_NeuroD', 'EBF1_ISL1',\n",
       "       'ISL1_Onecut2', 'LHX_NeuroD', 'EBF1_LHX', 'LHX_Onecut2',\n",
       "       'EBF1_NeuroD', 'NeuroD_Onecut2', 'EBF1_Onecut2', 'NKX2.5_NKX2.5',\n",
       "       'NKX2.5-alt_NKX2.5-alt', 'GATA_GATA', 'ISL1_ISL1', 'LHX_LHX',\n",
       "       'NeuroD_NeuroD', 'EBF1_EBF1', 'Onecut2_Onecut2'], dtype=object)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_all.motif_pair.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 36/36 [01:05<00:00,  1.81s/it]\n"
     ]
    }
   ],
   "source": [
    "import plotnine\n",
    "from plotnine import *\n",
    "plotnine.options.figure_size = (8, 2)\n",
    "\n",
    "for pair in tqdm(dfs_all.motif_pair.unique()):\n",
    "    df = dfs_all[dfs_all.motif_pair==pair]\n",
    "    df['profile/counts_max_ref_log2'] = np.log2(df['profile/counts_max_ref_frac'])\n",
    "    df['profile/counts_log2'] = np.log2(df['profile/counts_frac'])\n",
    "    \n",
    "    gmax = (ggplot(data = df, mapping = aes(x='distance', y='profile/counts_max_ref_log2')) + \n",
    "     geom_line(aes(color = 'task'))+ #, linetype = 'featured_task')) + \n",
    "     facet_wrap('~motif_pair_raw', nrow = 1) +\n",
    "     geom_hline(yintercept=0, alpha=0.2) + \n",
    "     scale_x_continuous(breaks = range(0,120,20), name = 'Center-to-center distance (bp)')+\n",
    "     scale_y_continuous(name = 'Max. log2(FC) of preds.')+\n",
    "     theme_minimal())\n",
    "    gmax.save(f'../../figures/{prefix}/8_in_silico_distances/max/insilico_perturb_max_{pair}.png', height = 2, width = 8)\n",
    "    gmax.save(f'../../figures/{prefix}/8_in_silico_distances/max/insilico_perturb_max_{pair}.pdf', height = 2, width = 8)\n",
    "    \n",
    "    gsum = (ggplot(data = df, mapping = aes(x='distance', y='profile/counts_log2')) + \n",
    "     geom_line(aes(color = 'task'))+ \n",
    "     facet_wrap('~motif_pair_raw', nrow = 1) +\n",
    "     geom_hline(yintercept=0, alpha=0.2) + \n",
    "     scale_x_continuous(breaks = range(0,120,20), name = 'Center-to-center distance (bp)')+\n",
    "     scale_y_continuous(name = 'Sum log2(fc) of preds.')+\n",
    "\n",
    "     theme_minimal())\n",
    "    gsum.save(f'../../figures/{prefix}/8_in_silico_distances/sum/insilico_perturb_sum_{pair}.png', height = 2, width = 8)\n",
    "    gsum.save(f'../../figures/{prefix}/8_in_silico_distances/sum/insilico_perturb_sum_{pair}.pdf', height = 2, width = 8)\n",
    "    gsum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bpnet-gpu",
   "language": "python",
   "name": "bpnet-gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
