---
title: 'Curation of CP and MN motifs'
author: "Melanie Weilert (1028-02098-001-001)"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    depth: 3
    theme: sandstone
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
  word_document:
    toc: yes
header-includes:
- \usepackage{fancyhdr}
- \usepackage{lipsum}
- \pagestyle{fancy}
- \fancyfoot[CO,CE]{1028-02098-001-001}
- \fancyfoot[LE,RO]{\thepage}
- \fancypagestyle{plain}{\pagestyle{fancy}}
editor_options:
  chunk_output_type: console
---

# Introduction

The goal of this analysis is to curate motifs found by the TF-MoDISco run such that there are no redundancies due palindromes, sequence similarity, overlapping peak regions etc. After this, we will begin to analyze, perturb, and interpret the motifs.

# Computational Setup

```{r, warning=F, message=F}
#Standard packages
library(rtracklayer) ; library(GenomicRanges); library(magrittr) ; library(Biostrings)
library(ggplot2) ; library(reshape2); library(plyranges); library(Rsamtools)
library(dplyr); library(data.table); library(patchwork); library(readr); library(testit)

#KNITR Options
setwd("/n/projects/mw2098/publications/2022_maven_ISL1/analysis/")
options(knitr.figure_dir="../figures/3_motif_curation", java.parameters = "- Xmx6g")

#Lab sources
source("../scripts/granges_common.r")
source("../scripts/metapeak_common.r")
source("../scripts/knitr_common.r")
source("../scripts/caching.r")
source("../scripts/metapeak_functions.R")

#Specific sources
library(BSgenome.Hsapiens.UCSC.hg19)
library(ggpubr)
source("../scripts/motif_functions.r")

#Pre-existing variables
model_prefix<-'seq_width1000-lr0.001-lambda100-n_dil_layers9-conv_kernel_size7-tconv_kernel_size7-filters64'
modisco_dir <- 'modisco/seq_width1000-lr0.001-lambda100-n_dil_layers9-conv_kernel_size7-tconv_kernel_size7-filters64/'
tasks <- c('I_WT_D6CM','I_WT_S3MN')
cp_peak_path <- '../bed/I_wt_d6cm_consensus_filtered.bed'
threads <- 5
dir.create('tsv/mapped_motifs/', recursive = T)
dir.create('bed/mapped_motifs/', recursive = T)
```

# Summary of motifs of interest

A summary of the TF-MoDISco runs on profile are below:

+ D6CM task
    + NKX2.5 (D6CM)
        + pattern 0: canonical GCACTTAA
        + patterns 4 and 5 also show low affinity version of this.
    + GATA (D6CM)
        + pattern 1: canonical GATA motif (CTTATCT/GAGAAG)
        + pattern 2: identical GATA motif with varying flanks. Will be ignored since CWM-scanning will map both and their relative contributions show identical patterning relative to tasks.
    + ISL1 (D6CM)
        + pattern 3: ISL1 motif (CTAATGG)
        + Will be called ISL1-D6CM because it was found in the D6CM task.
+ S3MN task
    + ISL1-ATrich (S3MN)
        + pattern 0
        + AT-rich ISL1-like motif (TAATTNNATT)
    + NeuroD (S3MN)
        + pattern 1: CAGCTG
        + This is detected as AP-4, NeuroD, NHLH1, and HEN1 in MotifMap.
    + EBF1 (S3MN)
        + pattern 2: canonical EBF1 motif (CCCNNGGGA)
    + ISL1-double (S3MN)
        + pattern 3: 28bp separation
        + pattern 6: 9bp separation
        + pattern 7: 10bp separation
        + Contains 1 high affinity ISL1 and a LHX nearby.
    + Onecut2 (S3MN)
            + pattern 4: ATTGATT
            + Low affinity version as pattern 8.
    + NKX2.5-alt (S3MN)
        + pattern 5: alternative NKX motif with sequence GCACTTAG. The last 'G' appears to make it more important for S3MN.

```{r}
motifs_of_interest.df <- data.frame(
  name = c('NKX2.5', 'GATA', 'ISL1', 
           'LHX', 'LHX-ISL1-28', 'LHX-ISL1-10', 'LHX-ISL1-9', 
           'NeuroD', 'EBF1', 'Onecut2', 'NKX2.5-alt'),
  task = c('I_WT_D6CM', 'I_WT_D6CM', 'I_WT_D6CM',
           'I_WT_S3MN', 'I_WT_S3MN', 'I_WT_S3MN', 'I_WT_S3MN', 
           'I_WT_S3MN', 'I_WT_S3MN', 'I_WT_S3MN', 'I_WT_S3MN'),
  pattern = c(0, 1, 3, 
              0, 3, 6, 7,
              1, 2, 4, 5)
)
motifs_of_interest.df
```

# Collect motif instances

First, extract each mapped set of instances from each pattern to allocate the non-curated set of motifs of interest. We will take motifs from the counts contribution TF-MoDISco run because of the high counts correlation in the trained model and due to the fact that the ChIP-seq data is low resolution and the profile will not necessarily be higher resolution sequence information.

```{r, eval = F}
cmds.vec<-lapply(1:nrow(motifs_of_interest.df), function(x){
  info.df<-motifs_of_interest.df[x,]
  cmd<-paste0('bpnet cwm-scan --contrib-file preds/', model_prefix, 
              '/contrib.h5 --patterns metacluster_0/pattern_', 
              info.df$pattern, ' ', 
              modisco_dir, '/', info.df$task, 
              '_profile tsv/mapped_motifs/', 
              info.df$name, '_instances_noncurated_0based.tsv.gz')
  return(cmd)
}) %>% unlist
cmds.vec<-c('#!bin/bash', cmds.vec, '')

readr::write_lines(cmds.vec, file = '../tmp/cwm_scan_indiv_instances.sh', sep = '\n')
```

Run CWM-scanning to allocate.

```{bash, eval = F}
bash /n/projects/mw2098/collaboration/for_srivastava/tmp/cwm_scan_indiv_instances.sh
```

Next import motifs.

```{r}
motifs.df<-lapply(1:nrow(motifs_of_interest.df), function(x){
  info.df<-motifs_of_interest.df[x,]
  df<-readr::read_tsv(paste0('tsv/mapped_motifs/', info.df$name, '_instances_noncurated_0based.tsv.gz')) %>%
    dplyr::mutate(pattern_name = info.df$name,
                  modisco_task = info.df$task) 
}) %>% rbindlist(fill = TRUE)
```

Report number of motifs.

```{r}
table(motifs.df$pattern_name, motifs.df$example_interval_from_task) 
```

# Remove redundant palindromic mappings

Motifs that are palindromic will be mapped on the positive and negative strand with almost identical ranges. Here, we will remove redundantly mapped motifs and select the unique motif maps that have the highest importance score.

```{r}
#Remove palindromic motifs for each pattern.
motifs.gr<-lapply(motifs.df$pattern_name %>% unique, function(x){
  message(x)
  gr<-remove_palindromic_motifs_from_bpnet_instances(dfi = motifs.df, motif = x)
}) %>% GRangesList %>% unlist
```

# Remove redundantly mapped motifs due to overlapping peaks

When training with BPNet, peaks are found for each TF. If a genomic region has more than one peak, that region sequence is trained/mapped twice. This results in overreporting of motifs and redundancies in motif mapping. To filter these, we will resolve motifs by coordinate, strand, name, and select the motif that has the highest `contrib_weighted` score.

```{r}
motifs_unique.gr <- motifs.gr %>% 
  as.data.frame %>%
  #creating a proxy group_by due to bugs with dplyr's grouping at high values.
  dplyr::mutate(coord_id = paste(seqnames, start, end, strand, pattern_name, sep = '-')) %>% 
  dplyr::group_by(coord_id) %>%
  dplyr::slice_max(order_by = contrib_weighted, n = 1) %>%
  dplyr::ungroup() %>% 
  dplyr::mutate(example_idx_deprecated = example_idx,
                motif_unique_id = 1:nrow(.)) %>% #Mark an temporary identifier for subsequent filtering
  dplyr::select(-example_idx, -coord_id) %>% #Remove because it cannot be grouped by example_idx 
  GenomicRanges::makeGRangesFromDataFrame(df = ., keep.extra.columns = T, starts.in.df.are.0based = F)
  
assert("Filtering by peak redundancy did not work.", ((motifs_unique.gr %>% granges %>% unique %>% length) == (motifs_unique.gr  %>% length)))
```

# Measure %age of motifs across enhancer edges

```{r}
edge_motifs.df<-motifs_unique.gr %>% plyranges::filter(pattern_center <=100 | pattern_center>=900) %>% .$pattern_name %>% table %>% as.data.frame
total_motifs.df<-motifs_unique.gr$pattern_name %>% table %>% as.data.frame
total_motifs.df %>%
  dplyr::left_join(., edge_motifs.df, by = '.') %>%
  dplyr::rename(total_motif_count = Freq.x,edge_motif_count = Freq.y) %>%
  dplyr::mutate(perc_on_edge = round(edge_motif_count/total_motif_count, 2))

motifs_unique.gr<-motifs_unique.gr %>% plyranges::filter(!(pattern_center <=100 | pattern_center>=900))
```

Less than 1% of all motifs are on mapped BPNet window edges. However, we will filter out these motifs because of downstream perturbation predictions.

# Resolve/mark overlap of similar motifs

Many of the mapped motifs contain similarities such that they will map over the same sets of coordinate regions. To resolve these, we need to approach each comparison with a hierarchical, individualized regimen. Below, the steps will be listed for each group.

## Check for motif overlaps

Ensure that there isn't an overrepresentation of unexpected overlapping motif pairs. We expect some overlaps, but checking to make sure that our curations are consistent. Here, we will filter out the longer motifs because we know that those overlaps will be highly represented.

```{r}
overlapping_motifs.gr<- motifs_unique.gr %>% 
  GenomicRanges::reduce(ignore.strand = T) %>% 
  plyranges::mutate(motif_island_id = 1:length(.))
overlapping_motifs.ov<-findOverlaps(motifs_unique.gr, overlapping_motifs.gr, ignore.strand = T)
assert('IDs do not allow indexing shortcut.', 1:length(motifs_unique.gr)==overlapping_motifs.ov@from)
motifs_unique.gr$motif_island_id<-overlapping_motifs.ov@to

#Group motifs by their island ID and display frequency of motif contents
grouped_motif_content.df<-motifs_unique.gr %>% as.data.frame %>% 
  dplyr::group_by(motif_island_id) %>% 
  dplyr::summarize(motif_content = paste0(sort(pattern_name), collapse = '_'))

grouped_motif_content.df$motif_content %>% table() %>% sort(decreasing = T) %>% head(n=15)
```

Here we can see that of the top 15 "overlapped sets" we can see that only about 20% of ISL1 motifs occur by themselves and that many of the motifs "double up" as a NKX2.5, GATA, or both. This is presumably due to the ATTA component of ISL1. There are also a lot of other overlaps such as NKX variants with one another.

Because there are so many overlaps, we will mark which ISL1 motifs are overlapped with another motif, but we will not "filter" or "mark" these identities otherwise. When performing motif pair synergy analysis, these motifs will be automatically filtered out in both genomic and in silico testing. 

## Description of motif overlap strategy

Here, as we are not sure of the role that competitive motif sites play in our system, we do not want to filter out motifs that overlap as they may be regions of interest. Instead, we will demarcate genomic regions that have sufficient overlaps based on each individual situation that we anticipate to be of interest. This will allow filtering strategies to be applied in the future such that each individual analysis can be tailored to specific questions about affinity, competition, and representative motif behaviors.

# Mark overlaps

Mark the coordinates for all the ISL1 partners in a column for later storage.

```{r}
testit::assert(motifs_unique.gr$motif_unique_id %>% unique %>% length == motifs_unique.gr %>% length)

#Mark all overlaps and corresponding coordinates
ov_with_itself.df<-findOverlaps(motifs_unique.gr, motifs_unique.gr, ignore.strand = T) %>%
  as.data.frame %>%
  dplyr::filter(queryHits != subjectHits) %>%
  dplyr::mutate(subjectChrom = seqnames(motifs_unique.gr) %>% as.character %>% .[subjectHits],
                subjectStart = start(motifs_unique.gr)[subjectHits],
                subjectEnd = end(motifs_unique.gr)[subjectHits],
                subjectName = motifs_unique.gr$pattern_name[subjectHits],
                subjectCoords = paste0(subjectChrom, ':', subjectStart, '-', subjectEnd, '_', subjectName))

#Merge to 'query-based' for many subject hits
query_with_itself.df<-ov_with_itself.df %>%
  dplyr::group_by(queryHits) %>%
  dplyr::summarize(subjectCoordsAll=paste0(sort(subjectCoords), collapse = '/')) %>%
  dplyr::arrange(queryHits)

#Mark overlaps as a separate column.
motifs_unique.gr$overlap_information<-NA
motifs_unique.gr$overlap_information[query_with_itself.df$queryHits] <- query_with_itself.df$subjectCoordsAll
```

# Finish formatting motifs

```{r}
motifs_curated.gr<-motifs_unique.gr %>%
  plyranges::select(-motif_unique_id) %>%
  plyranges::mutate(motif_id = 1:length(.))
```

Plot a general summary of the motifs.

```{r CM_vs_MN_merged_bias_curated_motif_freq, fig.height = 3, fig.width=10}
ggplot(motifs_curated.gr %>% as.data.frame, aes(x = pattern_name))+
  geom_bar(aes(fill = factor(contrib_weighted_cat, levels = c('high','medium','low'))), color = 'black')+
  scale_fill_manual(values = rev(c('#83adb5', '#c7bbc9', '#5e3c58')), name = 'contrib')+
  scale_x_discrete(name = 'motifs')+
  ggtitle('CM_vs_MN_merged_bias curated motif frequencies')+
  theme_classic()
```

# Annotate motifs by their collected regional grouping

Because we took motifs based on their relative contribution and not the featured `example_idx` column, we need a new grouping annotation to define the motifs when we are conducting perturbations and other grouping systems. We will call this index `region_id`.

```{r}
islands.gr<-resize(motifs_curated.gr, 200, 'center') %>% 
  GenomicRanges::reduce(ignore.strand = T) %>% 
  plyranges::mutate(region_id = 0:(length(.)-1)) # Make region_id 0-based in order to be compatible with BPNet indexing properties.
ov<-findOverlaps(motifs_curated.gr, islands.gr, ignore.strand = T)
testit::assert((ov@from %>% unique %>% length) ==( ov@from %>% length))

#Show hoy many motifs there are per window
width(islands.gr) %>% summary
ov@to %>% table %>% table

#Assign region_id to these motifs
motifs_curated.gr$region_id<-islands.gr$region_id[ov@to]
motifs_curated.gr$region_start_1based<-islands.gr[ov@to] %>% resize(., 1000, 'center') %>% start()
motifs_curated.gr$region_end_1based<-islands.gr[ov@to] %>% resize(., 1000, 'center') %>% end()

#Assert that all motifs are incorporated.
testit::assert('All motifs have a correct region', all(overlapsAny(motifs_curated.gr, resize(islands.gr, 1000, 'center'))))

#Export regions in 0-based coordinates
islands.gr<-islands.gr %>% resize(., 1000, 'center') 
rtracklayer::export(islands.gr %>% plyranges::mutate(name = region_id), 'bed/mapped_motifs/all_grouped_regions_0based.bed', format = 'BED')
```

# Export curated motifs

In the subsequent .bed file it will be the `[task]_[motif_id]_[region_id]` naming convention.

```{r}
motifs_curated.df<-motifs_curated.gr %>% as.data.frame()
readr::write_tsv(motifs_curated.df, 'tsv/mapped_motifs/all_instances_curated_1based.tsv.gz')
rtracklayer::export(motifs_curated.gr %>% plyranges::mutate(name = paste0(pattern_name, '_', motif_id, '_', region_id)), 
                    'bed/mapped_motifs/all_instances_curated_0based.bed', format = 'BED')
print(length(motifs_curated.gr))
```

In conclusion, motifs were curated based on redundancies and saved to 1-based and 0-based coordinates. These will be the reference motifs that we base subsequent analysis on.

# Define motifs that overlap with CPs.

```{r}
motif_ids_overlapping_cps<-subsetByOverlaps(motifs_curated.gr, 
                                            rtracklayer::import(cp_peak_path) %>% resize(1000, 'center'), 
                                            ignore.strand = T)$motif_id
write(motif_ids_overlapping_cps, 'tsv/motif_ids_overlapping_cps.txt')
```

# Session Information

For the purposes of reproducibility, the R/Bioconductor session information is printed below:

```{r sessioninfo}
sessionInfo()
```












